### PR TITLE
Move xUnit tests in new folder

### PR DESCRIPTION
**Pull request by iSazonov**
(Replace #8237)

## Motivation

I have a PR where there are many new xUnit tests.
It would also be useful to create new xUnit tests for public APIs.
The number of xUnit tests will increase and their ordering is required.

## PR Summary

- Move C# xUnit tests in new folder. This allows to put new xUnit tests in directory structure in accordance with directory structure where cs files are.
- Use an xUnit TestCaseOrderer attribute to sequentially process tests for `powershell.config.json`.
- Update README.md
- A race condition was fixed which allowed  to run all XUnit tests in single batch job.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: test/xUnit/Asserts/PriorityOrderer.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+using System;
```
Comment (adityapatwardhan) [inline test/xUnit/Asserts/PriorityOrderer.cs]: Please add copyright header.

File: test/xUnit/Asserts/PriorityAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+using System;
```
Comment (adityapatwardhan) [inline test/xUnit/Asserts/PriorityAttribute.cs]: Please add copyright header.

File: build.psm1 [hunk 3]
```diff
@@ -1419,30 +1417,22 @@ function Start-PSxUnit {
...
-        Publish-TestResults -Path $SequentialTestResultsFile -Type 'XUnit' -Title 'Xunit Sequential'
+        dotnet build --configuration $Options.configuration 
 
-        $extraParams = @()
+        if (Test-Path $ParallelTestResultsFile) {
+            Remove-Item $ParallelTestResultsFile -Force -ErrorAction SilentlyContinue
+        }
 
         # we are having intermittent issues on macOS with these tests failing.
         # VSTS has suggested forcing them to be sequential
```
Comment (adityapatwardhan) [inline build.psm1]: Can we validate if this is still the case. If not, we can removed the condition.

File: test/xUnit/csharp/test_Runspace.cs [hunk 4]
```diff
@@ -66,7 +66,7 @@ public void TestRunspaceWithPowerShell()
         [Fact]
         public void TestRunspaceWithPowerShellAndInitialSessionState()
         {
-            InitialSessionState iss = InitialSessionState.CreateDefault2();
+            InitialSessionState iss = InitialSessionState.CreateDefault();
```
Comment (adityapatwardhan) [inline test/xUnit/csharp/test_Runspace.cs]: Why was this changed?

**Review by adityapatwardhan — DISMISSED** [review]

File: test/xUnit/Asserts/PriorityOrderer.cs (hunk 1 above)
Comment (iSazonov) [reply test/xUnit/Asserts/PriorityOrderer.cs]: Fixed.

File: test/xUnit/Asserts/PriorityAttribute.cs (hunk 2 above)
Comment (iSazonov) [reply test/xUnit/Asserts/PriorityAttribute.cs]: Fixed.

File: build.psm1 (hunk 3 above)
Comment (iSazonov) [reply build.psm1]: Perhap @TravisEz13 can comment this.

File: test/xUnit/csharp/test_Runspace.cs (hunk 4 above)
Comment (iSazonov) [reply test/xUnit/csharp/test_Runspace.cs]: Two tests above use the `CreateDefault()`. I guess it was typo. If we need tests for `CreateDefault2()` I believe it should be added in new PR.

Comment (adityapatwardhan) [reply test/xUnit/csharp/test_Runspace.cs]: I believe `CreateDefault2` is intentional as it creates an ISS without any modules loaded. Please change it back.

File: build.psm1 (hunk 3 above)
Comment (TravisEz13) [reply build.psm1]: We would have to try it to know.

File: test/xUnit/csharp/test_Runspace.cs (hunk 4 above)
Comment (iSazonov) [reply test/xUnit/csharp/test_Runspace.cs]: Reverted with new comment.

File: build.psm1 (hunk 3 above)
Comment (iSazonov) [reply build.psm1]: We could open a tracking issue for the work and don't block the PR.

Also we have many CodeFactor issues. Also we need to have more tests for binary APIs. Also we need to have o\performance tests. I think we should open new tracking issues.

File: test/xUnit/Asserts/PriorityAttribute.cs (hunk 2 above)
Comment (adityapatwardhan) [reply test/xUnit/Asserts/PriorityAttribute.cs]: Still missing copyright header here.

Comment (iSazonov) [reply test/xUnit/Asserts/PriorityAttribute.cs]: Sorry, I lost new commits.

Fixed.

File: build.psm1 (hunk 3 above)
Comment (iSazonov) [reply build.psm1]: Done #8400

**Comment by adityapatwardhan** [comment]
@iSazonov Please resolve the merge conflict.

File: test/xUnit/xUnit.tests.csproj [hunk 5]
```diff
@@ -5,6 +5,7 @@
   <PropertyGroup>
     <Description>PowerShell xUnit Tests</Description>
     <AssemblyName>powershell-tests</AssemblyName>
+    <GenerateProgramFile>true</GenerateProgramFile>
```
Comment (adityapatwardhan) [inline test/xUnit/xUnit.tests.csproj]: Why was this added?

Comment (iSazonov) [reply test/xUnit/xUnit.tests.csproj]: It was temporary additions.

Removed.

**Comment by iSazonov** [comment]
I updated #8400 to fix the rare race condition in tests.

@adityapatwardhan We could merge.

**Comment by iSazonov** [comment]
@adityapatwardhan Could you please update your code review and merge?

**Comment by TravisEz13** [comment]
@iSazonov 	@adityapatwardhan 	Is out of the office until the new year.

**Comment by iSazonov** [comment]
@TravisEz13 If you could merge I could continue my work.

**Comment by TravisEz13** [comment]
rebased the branch mainly to verify the tests still pass

File: test/xUnit/csharp/test_Runspace.cs (hunk 4 above)
Comment (vexx32) [reply test/xUnit/csharp/test_Runspace.cs]: @adityapatwardhan that seems like a method name that probably needs revising, if we can, or adding an alias for in some fashion. Ilya is generally pretty familiar with the repo; if he got it confused, it's likely others will also.

Comment (TravisEz13) [reply test/xUnit/csharp/test_Runspace.cs]: Changing the methodName would be a huge breaking change, but we could add a new method that calls this that is named better.  I would suggest filing an issue about this.

**Review by TravisEz13 — APPROVED** [review]
